pkg/controller/v1beta1/inferenceservice/reconcilers/hpa/hpa_reconciler.go

```diff
@@ -140,11 +140,28 @@ func (r *HPAReconciler) checkHPAExist(client client.Client) (constants.CheckResu
 	if semanticHPAEquals(r.HPA, existingHPA) {
 		return constants.CheckResultExisted, existingHPA, nil
 	}
+	if shouldDeleteHPA(r.HPA) {
+		return constants.CheckResultDelete, existingHPA, nil
+	}
 	return constants.CheckResultUpdate, existingHPA, nil
 }
 
 func semanticHPAEquals(desired, existing *autoscalingv2.HorizontalPodAutoscaler) bool {
-	return equality.Semantic.DeepEqual(desired.Spec, existing.Spec)
+	desiredAutoscalerClass, hasDesiredAutoscalerClass := desired.Annotations[constants.AutoscalerClass]
+	existingAutoscalerClass, hasExistingAutoscalerClass := existing.Annotations[constants.AutoscalerClass]
+	var autoscalerClassChanged bool
+	if hasDesiredAutoscalerClass && hasExistingAutoscalerClass {
+		autoscalerClassChanged = desiredAutoscalerClass != existingAutoscalerClass
+	} else if hasDesiredAutoscalerClass || hasExistingAutoscalerClass {
+		autoscalerClassChanged = true
+	}
+
+	return equality.Semantic.DeepEqual(desired.Spec, existing.Spec) && !autoscalerClassChanged
+}
+
+func shouldDeleteHPA(desired *autoscalingv2.HorizontalPodAutoscaler) bool {
```

Comment:
I think we should only delete if it is changed from enabled to external otherwise delete throws error

Reply:
Actually nvm, `shouldDeleteHPA` is called only when HPA exists

Reply:
have you tested this? I think it will not work

Reply:
@yuzisun Please take another look. I tested a couple of scenarios and they worked as expected.